# 예외 처리하기, try-catch문

```java

try{
  // 예외가 발생할 가능성이 있는 문장들을 넣는다
} catch (Exception e) {
  // Exception1이 발생했을 경우, 이를 처리하기 위한 문장을 적는다
}

```

Exception 은 모든 예외의 최고조상이다. 따라서 모든 예외처리가 가능하다

# 프로그램 오류

- 컴파일 에러 : 컴파일 할 때 발생하는 에러
- 런타임 에러 : 실행 할 때 발생하는 에러
- 논리적 에러 : 작성 의도와 다르게 동작

# printStackTrace()와 getMessage()

예외가 발생되면 예외 객체가 생성된다. 그 때 메서드들을 통해 오류에 대한 정보 알 수 있음

- printStackTrace() : 예외발생 당시의 호출스택에 있었던 메서드의 정보와 예외 메시지를 화면에 출력
- getMessage() : 발생한 예외클래스의 인스턴스에 정장된 메시지를 얻을 수 있다

# 멀티 catch 블럭

- 내용이 같은 catch블럭을 하나로 합친 것

```java

try {
  ...
} catch(ExceptionA e){

} catch(ExceptionB e2){

}

```

```java

try{
  ...
} catch (ExceptionA | ExceptionB e){

}

```

부모 자식 관계일 때는 부모만 써주면 된다

# 예외 발생시키기

1. 연산자 new를 이용해서 발생시키려는 예외 클래스의 객체를 만든다

Exception e = new Exception("");

2. 키워드 throw를 이용해서 예외를 발생시킨다

throw e;

# checked 예외, unchecked 예외

- checked 예외 : 컴파일러가 예외처리 여부를 체크(예외 처리 필수)
- unchecked 예외 : 컴파일러가 예외 처리 여부를 체크 안함(예외 처리 선택)

Exception과 그 자손은 반드시 예외처리를 해야한다

RuntimeException과 그 자손은 예외처리를 해주지 않아도 컴파일이 된다

